### ISSUE TITLE
New moment definition (oct. 2018) looks like limitate moment usage not necessarily

### ISSUE THREAD

**Issue by pieplu** [flow-typed/flow-typed#2867]
From comment on https://github.com/flow-typed/flow-typed/pull/2819

This update trigger new flow errors, and I think it's a problem:

Example:
```
type itemType = {
    maybeDate?: ?Date,
    otherMaybeDate?: ?Date
}
const item:  ={
    maybeDate: undefined,
    otherMaybeDate: null
};
moment(item.maybeDate); // flow error but valid moment result (moment.now()
moment(item.otherMaybeDate); // flow error but moment result (unvalid date, but a moment object usable)
```
Output extract:
```
11:08:16     21|     maybeDate?: ?Date,
11:08:16                                     ^^^^^ [1]
11:08:16    flow-typed/npm/moment_v2.x.x.js:128:19
11:08:16    128|   static (string: string, format: string | Array<string>): moment$Moment;
11:08:16                           ^^^^^^ [2]
11:08:16    flow-typed/npm/moment_v2.x.x.js:130:13
11:08:16    130|     string: string,
11:08:16                     ^^^^^^ [3]
11:08:16    flow-typed/npm/moment_v2.x.x.js:135:13
11:08:16    135|     string: string,
11:08:16                     ^^^^^^ [4]
11:08:16    flow-typed/npm/moment_v2.x.x.js:140:13
11:08:16    140|     string: string,
11:08:16                     ^^^^^^ [5]
```

This it introduce a limitation to the moment library usage, than moment do not exclude. I will open a issue

I think the main problem is that this code trigger error: `moment(undefined)`

_Originally posted by @pieplu in https://github.com/flow-typed/flow-typed/pull/2819/comment#issuecomment-431871621_

**Comment by valscion** [comment]
Hi, and thanks for opening this issue :relaxed:

> I think the main problem is that this code trigger error: `moment(undefined)`

Yes, that is the main problem.

I can see how that generates a valid moment object. However, this feature of `moment` is not documented here: https://momentjs.com/docs/#/parsing/

As this is not documented, I don't think the types should allow `undefined` to be passed in, and I think the types before #2819 were in fact broken. 

---

**Try to let me explain my thoughts:**

To take your example:

```js
type itemType = {
    maybeDate?: ?Date,
    otherMaybeDate?: ?Date
}
const item: itemType  ={
    maybeDate: undefined,
    otherMaybeDate: null
};

moment(item.maybeDate); // case A
moment(item.otherMaybeDate); // case B
```

I think it is surprising that the _case A_ uses the current time when `maybeDate` is missing. Explicitly checking the value presence makes the code less surprising:

```js
// case A fixed
if (item.maybeDate) {
  moment(item.maybeDate);
} else {
  // Use current time if maybeDate is missing
  moment();
}
```

For the _case B_, as you mentioned, it outputs an invalid moment object as `null` is passed to the constructor. So the type error in there is a good thing, as it protects you from invalid usage. So for B, you will still need to have this conditional wrapper:

```js
// case B fixed
if (item.otherMaybeDate) {
  moment(item.otherMaybeDate);
} else {
  // Use current time if otherMaybeDate is missing
  moment();
}
```

---

**TL:DR;**
I think the type error you are seeing is as expected, and is not a bug in the type definitions. Instead, the changes done in #2819 fixed the types as they were too lax.

Ping @GAntoine

**Comment by valscion** [comment]
> On Moment typescript, they allow "void", with the comment `// null, undefined`.
> https://github.com/moment/moment/blob/develop/moment.d.ts#L400
> 
> It's looks like null and undefined are alowed, ( void is added in this commit [moment/moment@5a1843d](https://github.com/moment/moment/commit/5a1843de5a98025aa04437df8d7e4e7200306704) ) and void allow null and undefined in typescript ( https://www.typescriptlang.org/docs/handbook/basic-types.html#void )

Thanks for pointing me to this direction! Looks like `moment` project does allow `null` and `undefined` to be passed in, so we should, too. I'll prepare a PR shortly to do this.
